File: src/haskell/utf-256.hs
{-# LANGUAGE OverloadedStrings #-}

import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BSC
import Data.Word (Word8)
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Data.Bits (shiftR, (.&.), shiftL, (.|.))
import Numeric (showHex)
import Control.Monad (when)

chunkSize :: Int
chunkSize = 8

printHelp :: String -> IO ()
printHelp progname = do
    hPutStrLn stderr ""
    hPutStrLn stderr "UTF-256 Encoder/Decoder"
    hPutStrLn stderr ""
    hPutStrLn stderr "Usage:"
    hPutStrLn stderr ("  " ++ progname ++ " -e UTF-8_FILE -o OUTPUT_FILE")
    hPutStrLn stderr ("  " ++ progname ++ " -d UTF-256_FILE -o OUTPUT_FILE")
    hPutStrLn stderr ""

encode :: FilePath -> FilePath -> IO Int
encode infile outfile = do
    content <- BS.readFile infile
    let encoded = BS.concatMap encodeByte content
    BS.writeFile outfile encoded
    return 0
  where
    encodeByte :: Word8 -> BS.ByteString
    encodeByte byte = BS.pack [if testBit byte i then 0xFF else 0x00 | i <- [7,6..0]]
    testBit b i = (b `shiftR` i) .&. 1 == 1

decode :: FilePath -> FilePath -> IO Int
decode infile outfile = do
    content <- BS.readFile infile
    if BS.length content `mod` chunkSize /= 0
      then do
        hPutStrLn stderr "Error: file is not a multiple of 8 bytes"
        return 1
      else do
        let chunks = chunkBS chunkSize content
        decodedEither <- mapM decodeChunk chunks
        case sequence decodedEither of
          Left err -> do
            hPutStrLn stderr err
            return 1
          Right decoded -> do
            BS.writeFile outfile (BS.pack decoded)
            return 0

  where
    chunkBS :: Int -> BS.ByteString -> [BS.ByteString]
    chunkBS n bs
      | BS.null bs = []
      | otherwise = let (h, t) = BS.splitAt n bs in h : chunkBS n t

    decodeChunk :: BS.ByteString -> IO (Either String Word8)
    decodeChunk chunk = do
        let bytes = BS.unpack chunk
        decodeBits bytes 0

    decodeBits :: [Word8] -> Word8 -> IO (Either String Word8)
    decodeBits [] acc = return $ Right acc
    decodeBits (b:bs) acc
      | b == 0x00 = decodeBits bs (acc `shiftL` 1)
      | b == 0xFF = decodeBits bs ((acc `shiftL` 1) .|. 1)
      | otherwise = return $ Left ("Invalid byte in UTF-256 file: 0x" ++ showHex b "")

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName

    if length args < 1
      then printHelp progname >> exitFailure
      else parseArgs args Nothing Nothing Nothing

  where
    parseArgs :: [String] -> Maybe FilePath -> Maybe FilePath -> Maybe FilePath -> IO ()
    parseArgs [] (Just encodeFile) Nothing (Just outputFile) = do
        r <- encode encodeFile outputFile
        when (r == 0) $ putStrLn $ "Encoded successfully to '" ++ outputFile ++ "'"
        if r /= 0 then exitFailure else return ()

    parseArgs [] Nothing (Just decodeFile) (Just outputFile) = do
        r <- decode decodeFile outputFile
        when (r == 0) $ putStrLn $ "Decoded successfully to '" ++ outputFile ++ "'"
        if r /= 0 then exitFailure else return ()

    parseArgs [] _ _ _ = do
        hPutStrLn stderr "Error: Invalid or missing arguments."
        printHelp =<< getProgName
        exitFailure

    parseArgs ("-e":file:rest) Nothing Nothing output = parseArgs rest (Just file) Nothing output
    parseArgs ("-d":file:rest) encode Nothing output = parseArgs rest encode (Just file) output
    parseArgs ("-o":file:rest) encode decode Nothing = parseArgs rest encode decode (Just file)

    parseArgs (arg:_) _ _ _ = do
        hPutStrLn stderr $ "Unknown or malformed argument: " ++ arg
        printHelp =<< getProgName
        exitFailure

File: src/java/utf256.java
import java.io.*;
import java.nio.file.Files;

public class utf256 {

    private static final int CHUNK_SIZE = 8;

    public static void printHelp(String progName) {
        System.err.println("\nUTF-256 Encoder/Decoder\n");
        System.err.println("Usage:");
        System.err.println("  " + progName + " -e UTF-8_FILE -o OUTPUT_FILE");
        System.err.println("  " + progName + " -d UTF-256_FILE -o OUTPUT_FILE\n");
    }

    public static int encode(String infile, String outfile) {
        try (InputStream in = new BufferedInputStream(new FileInputStream(infile));
             OutputStream out = new BufferedOutputStream(new FileOutputStream(outfile))) {

            int byteRead;
            while ((byteRead = in.read()) != -1) {
                for (int i = 7; i >= 0; i--) {
                    int bit = (byteRead >> i) & 1;
                    out.write(bit == 1 ? 0xFF : 0x00);
                }
            }
        } catch (IOException e) {
            System.err.println("Error during encoding: " + e.getMessage());
            return 1;
        }
        return 0;
    }

    public static int decode(String infile, String outfile) {
        try (InputStream in = new BufferedInputStream(new FileInputStream(infile));
             OutputStream out = new BufferedOutputStream(new FileOutputStream(outfile))) {

            byte[] chunk = new byte[CHUNK_SIZE];
            int bytesRead;
            while ((bytesRead = in.read(chunk)) == CHUNK_SIZE) {
                int decodedByte = 0;
                for (int i = 0; i < CHUNK_SIZE; i++) {
                    byte b = chunk[i];
                    if (b == 0x00) {
                        decodedByte <<= 1;
                    } else if ((b & 0xFF) == 0xFF) {
                        decodedByte = (decodedByte << 1) | 1;
                    } else {
                        System.err.printf("Invalid byte in UTF-256 file: 0x%02X%n", b);
                        return 1;
                    }
                }
                out.write(decodedByte);
            }
            if (bytesRead != -1) {
                System.err.println("Error: file is not a multiple of 8 bytes");
                return 1;
            }
        } catch (IOException e) {
            System.err.println("Error during decoding: " + e.getMessage());
            return 1;
        }
        return 0;
    }

    public static void main(String[] args) {
        String encodeFile = null;
        String decodeFile = null;
        String outputFile = null;

        if (args.length < 1) {
            printHelp("Utf256");
            System.exit(1);
        }

        // Simple flag parsing
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-e":
                    if (i + 1 < args.length) {
                        encodeFile = args[++i];
                    } else {
                        printHelp("Utf256");
                        System.exit(1);
                    }
                    break;
                case "-d":
                    if (i + 1 < args.length) {
                        decodeFile = args[++i];
                    } else {
                        printHelp("Utf256");
                        System.exit(1);
                    }
                    break;
                case "-o":
                    if (i + 1 < args.length) {
                        outputFile = args[++i];
                    } else {
                        printHelp("Utf256");
                        System.exit(1);
                    }
                    break;
                default:
                    printHelp("Utf256");
                    System.exit(1);
            }
        }

        if ((encodeFile != null && decodeFile != null) || (encodeFile == null && decodeFile == null)) {
            System.err.println("Error: You must specify either -e or -d, but not both.");
            printHelp("Utf256");
            System.exit(1);
        }

        if (outputFile == null) {
            System.err.println("Error: Output file is required (-o).");
            printHelp("Utf256");
            System.exit(1);
        }

        int result = 0;
        if (encodeFile != null) {
            result = encode(encodeFile, outputFile);
            if (result == 0) {
                System.out.println("Encoded successfully to '" + outputFile + "'");
            }
        } else if (decodeFile != null) {
            result = decode(decodeFile, outputFile);
            if (result == 0) {
                System.out.println("Decoded successfully to '" + outputFile + "'");
            }
        }

        System.exit(result);
    }
}
